### ISSUE TITLE
Streaming K-Means?

### ISSUE THREAD

**Issue by yasab27** [facebookresearch/faiss#4499]
Hey folks, 

I apologize if this is not the correct place to ask this question, but I wanted to reach out only guidance to go about implementing a streaming K-means using FAISS. I am new to the world of highly scaled vector search so forgive me if this is elementary. 

Basically, I am interested in computing centroids of a dataset of roughly ~1 billion dense vectors, ideally on GPU. These are all in FP16 but cannot fit in memory (~4 TB). What is the best pattern/protocol to implement a streaming version of this algorithm? I currently have an implementation where I stream in ~1 million vectors compute the centroids ( K ~ O(1000) - O(10,000) ) and add them to the Faiss Flat2 Index. Every full batch through the data I just basically store in CPU RAM the assignments and sum of square distances of all assigned query vectors (i.e. my data) against a fixed index of centroids (initialized randomly), and then update the centroids in the Flat Index after batching through the entire dataset and then repeating this over several epochs. 

I'm not sure if there a is a better strategy for this or if I'm really exploiting FAISS to its best use, the major advantage I think I get from this protocol is just fast querying against the centroid index. 

My fundamental question is whether there is a better strategy for doing this or if the team has plans down the road to support a first party streaming API. 

Thanks!

Yasa 
Stanford PhD Student

**Comment by satymish** [comment]
Can you clarify is this is what you're doing?
1. You compute centroids for the first batch of vectors(1 million)
2. You find assignments for all 1 billion vectors to these centroids
3. Then you recalculate centroids for the second batch of vectors(next 1 million) using the previous centroids as the initial centroids
4. repeat steps 2-3 until you run through the whole dataset

**Comment by onestardao** [comment]
interesting line of thinking, and i think you're asking a very real question hidden under it

streaming k-means over faiss vectors is tricky because most faiss clustering logic assumes full-batch access at index build time

your sketch suggests doing partial updates over batches, and carrying centroids forward — which feels like online k-means or reservoir-style memory compression

one thing that breaks easily in that setup is centroid drift, especially if the early batches are biased or not representative

you might want to check if the faiss Clustering class allows centroid injection between calls, or if you're forced to restart from scratch each time

another route is to design a memory-efficient index like ivf-hnsw that allows staged updates, and then fit a meta clustering pass only over visited centroids

this is a hard problem but definitely worth solving, especially in long-horizon LLM systems where memory grows over time

**Comment by yasab27** [comment]
I realize my initial question was a bit vague, here is a more explicit breakdown of what I am proposing: 

#### Algorithm Overview

1. **Initialise centroids**
   * Select **K** starting centroids (e.g. uniform random, k-means++, or sample of the data).
   * Build a GPU Faiss `IndexFlatL2` (or `IndexFlatIP`) containing these K vectors.

2. **E-step (streaming)**
   * For each disk batch (≈ 10⁶ vectors):
     1. Copy the batch to GPU.
     2. Query the frozen index for the nearest centroid of every vector.
     3. Accumulate per-centroid statistics on-GPU  
        * `n_j` – number of assignments  
        * `s_j` – element-wise sum of assigned vectors  
   * Only **O(K × d)** floats and **O(K)** ints are stored, irrespective of dataset size.

3. **M-step**
   * For every centroid `j` with `n_j > 0` update  
     
       c_j <- s_j/n_j
     
   * Rebuild the Faiss index with the new centroids (or in-place overwrite + `index.reset(); index.add()`).

4. **Repeat E- & M-steps** until convergence (fixed epochs, centroid shift < ε, or stabilised objective).

Because counts and sums are *additive*, batching is mathematically exact: the result equals a full-dataset pass in classic Lloyd k-means, but with drastically lower memory pressure.

---

#### Pseudocode

```python
# Inputs: dataset_on_disk, batch_size, K, max_epochs
centroids = init_centroids(K)                    # Step 1
index = faiss.IndexFlatL2(d)
index.add(centroids)

for epoch in range(max_epochs):

    # reset running statistics
    n   = np.zeros(K, dtype=np.int64)
    s   = np.zeros((K, d), dtype=np.float32)

    # ---------- E-step ----------
    for X_b in stream_batches(dataset_on_disk, batch_size):
        X_gpu = to_gpu(X_b)
        _, assign = index.search(X_gpu, 1)       # nearest centroid IDs

        # accumulate per-centroid stats on GPU, then reduce
        for j in range(K):
            mask       = (assign == j)
            n[j]      += mask.sum()
            s[j, :]   += X_gpu[mask].sum(axis=0)
    # ----------------------------

    # ---------- M-step ----------
    for j in range(K):
        if n[j] > 0:
            centroids[j] = s[j] / n[j]
        else:                                     # handle empty cluster
            centroids[j] = reseed()               # e.g. random data point

    index.reset()                                 # rebuild flat index
    index.add(centroids)
    # ----------------------------

    if converged(previous_centroids, centroids):
        break
```

**Comment by yasab27** [comment]
The major gain here I think is just leveraging FAISS for the super fast centroid lookup, i.e. minimizing the per batch latency.

**Comment by onestardao** [comment]
Thanks for the detailed breakdown ~ this is a great setup, and it’s clearer now where the pressure points are.

You’re absolutely right that FAISS helps here with low-latency centroid lookup. The tricky part, though, is what happens across batches ~  centroid drift, stale accumulations, or non-representative early samples can quietly poison the convergence.

I recently published a detailed breakdown on common failure modes in vector architecture, including scenarios like this. You might find this angle helpful as well:

https://github.com/onestardao/WFGY/blob/main/ProblemMap/vectorstore-metrics-and-faiss-pitfalls.md

Happy to go deeper if you want to run through specific cases or test edge cases.
Let’s crack this thing properly.

**Comment by mdouze** [comment]
The reason why Faiss does not have a streaming k-means implementation is because it is not clear when it is useful. 
In terms of k (nb centroids) and n (nb training vectors), we are in one of the following cases: 

1. n, k are small enough to run confortably
2. n is very large (say 1B) and k is small (say 100k) --> then you should use subsampling to get fewer training vectors per centroid, see https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means 
3. n is very large (say 1B) and k is large (say 10M) --> then the problems are (1) storing the training set and (2) computational cost of clustering --> a distribtuted k-means implementation is needed, see  https://github.com/facebookresearch/faiss/tree/main/benchs/distributed_ondisk
4. k is larger than n --> then there is not much we can do...
